include/bm/bm_sim/periodic_task.h

```diff
@@ -0,0 +1,92 @@
+/* Copyright 2013-present Barefoot Networks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef BM_BM_SIM_PERIODIC_TASK_H_
+#define BM_BM_SIM_PERIODIC_TASK_H_
+
+#include <string>
+#include <vector>
+#include <thread>
+#include <atomic>
+#include <queue>
+#include <chrono>
+#include <functional>
+#include <condition_variable>
+#include <mutex>
+
+struct PeriodicTask {
+  void cancel();
+  void reset_next();
+
+  const std::string name;
+  const std::function<void()> fn;
+  const std::chrono::milliseconds interval;
+  std::chrono::system_clock::time_point next;
+
+  PeriodicTask(const std::string &name,
+               std::function<void()> fn,
+               std::chrono::milliseconds interval);
+
+  ~PeriodicTask();
+};
+
+// This smells a bit (binding 'this' in inline member initialization),
+// but it compiles and I haven't found documentation against it
+#define REGISTER_EXTERN_PERIODIC(extrn_name, fn, interval) \
+  PeriodicTask fn## __task = \
+      PeriodicTask(#extrn_name #fn, \
+                   std::bind(&extrn_name::fn, this), \
+                   interval);
+
+#define REGISTER_GLOBAL_PERIODIC(fn, interval) \
+  static PeriodicTask fn## __task = PeriodicTask(#fn, fn, interval);
+
+class PeriodicTaskList {
+  std::mutex queue_mutex;
+  std::condition_variable cv;
+
+  class PeriodCompare {
+   public:
+    bool
+    operator() (const PeriodicTask *lhs, const PeriodicTask *rhs) {
+      return lhs->next > rhs->next;
+    }
+  };
+
+  using TaskQueue = std::priority_queue<PeriodicTask*,
+                                        std::vector<PeriodicTask*>,
+                                        PeriodCompare>;
+
+  TaskQueue task_queue;
+
+  std::atomic<bool> running;
```

Comment:
Not sure why you need 2 variables (`running` and `exiting`)
Also since you have a mutex anyway, I don't think there is a big advantage in making them atomic rather than just using the mutex.